salt/modules/nxos.py

```diff
@@ -1,47 +1,208 @@
+
+    The nxos.cmd <function> syntax is preserved for backwards compatibility.
 '''
-# Import Python libs
 from __future__ import absolute_import, print_function, unicode_literals
 
+# Import python stdlib
+import logging
+import difflib
+import re
+import ast
+
 # Import Salt libs
+from salt.utils.pycrypto import gen_hash, secure_password
 import salt.utils.platform
+import salt.utils.nxos
+from salt.ext import six
+from salt.exceptions import CommandExecutionError, NxosError
+from socket import error as socket_error
 
-__proxyenabled__ = ['nxos']
 __virtualname__ = 'nxos'
 
+log = logging.getLogger(__name__)
+
+DEVICE_DETAILS = {'grains_cache': {}}
+COPY_RS = 'copy running-config startup-config'
+
+CONNECTION_ERROR_MSG = '''
+    Unable to send command to the NX-OS device.
+    Please verify the following and re-try:
+    - 'feature ssh' must be enabled for SSH proxy minions.
+    - 'feature nxapi' must be enabled for NX-API proxy minions.
+    - Settings in the proxy minion configuration file must match device settings.
+    - NX-OS device is reachable from the Salt Master.
+'''
+
 
 def __virtual__():
+    return __virtualname__
+
+
+def ping(**kwargs):
+    '''
+    Ping the device on the other end of the connection.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd ping
+    '''
     if salt.utils.platform.is_proxy():
-        return __virtualname__
-    return (False, 'The nxos execution module failed to load: '
-            'only available on proxy minions.')
+        return __proxy__['nxos.ping']()
+    return __utils__['nxos.ping'](**kwargs)
+
+
+# -----------------------------------------------------------------------------
+# Device Get Functions
+# -----------------------------------------------------------------------------
+def check_password(username, password, encrypted=False, **kwargs):
+    '''
+    Verify user password.
+
+    username
+        Username on which to perform password check
+
+    password
+        Password to check
+
+    encrypted
+        Whether or not the password is encrypted
+        Default: False
+
+    .. code-block: bash
 
+        salt '*' nxos.cmd check_password username=admin password=admin
+        salt '*' nxos.cmd check_password username=admin \\
+            password='$5$2fWwO2vK$s7.Hr3YltMNHuhywQQ3nfOd.gAPHgs3SOBYYdGT3E.A' \\
+            encrypted=True
+    '''
+    hash_algorithms = {'1': 'md5',
+                       '2a': 'blowfish',
+                       '5': 'sha256',
+                       '6': 'sha512', }
+    password_line = get_user(username, **kwargs)
+    if not password_line:
+        return None
+    if '!' in password_line:
+        return False
+    cur_hash = re.search(r'(\$[0-6](?:\$[^$ ]+)+)', password_line).group(0)
+    if encrypted is False:
+        hash_type, cur_salt, hashed_pass = re.search(r'^\$([0-6])\$([^$]+)\$(.*)$', cur_hash).groups()
+        new_hash = gen_hash(crypt_salt=cur_salt, password=password, algorithm=hash_algorithms[hash_type])
```

Comment:
gen_hash uses python's `crypt.crypt`, which says:

> This module implements an interface to the crypt(3) routine, which is a one-way hash function based upon a modified DES algorithm; see the Unix man page for further details. Possible uses include storing hashed passwords so you can check passwords without storing the actual password, or attempting to crack Unix passwords with a dictionary.
>
> Notice that the behavior of this module depends on the actual implementation of the crypt(3) routine in the running system. Therefore, any extensions available on the current implementation will also be available on this module.

[Looking on wikipedia, I see this](https://en.wikipedia.org/wiki/Crypt_(C)#Support_in_operating_systems):
![image](https://user-images.githubusercontent.com/111300/78938878-43c2ff00-7a70-11ea-8f51-cd5ac82a1d2b.png)

I need to modify `gen_hash` so that it reflects the crypt methods available on the current system.

I'll also need to modify `set_password` to make sure that the hash method requested by the user is available on NXOS. I would prefer to raise an error reporting the problem to the user rather than silently setting a hash not usable on NXOS or lowering the hash method to the lowest common denominator.

I did a quick search, and it appears that base OS for NXOS is linux, but your list of hash_algorithms includes 2a - which isn't available on my Linux test box.

@mikewiebe, can you please verify what password hash algorithms are supported on nxos? If python can run, [`crypt.methods` will show the list of supported algorithms](https://docs.python.org/3/library/crypt.html#crypt.methods).

Reply:
I've done the gen_hash fixes and am planning on restricting the usable hashes to those normally available on linux.

Reply:
@dhiltonp Here is the list of supported algorithms on nxos.

```python
>>> import crypt
>>> crypt.methods
[<crypt.METHOD_SHA512>, <crypt.METHOD_SHA256>, <crypt.METHOD_MD5>, <crypt.METHOD_CRYPT>]
>>>
```

Reply:
@dhiltonp I should point out that much of this functionality existed before my changes to combine (ssh and nxapi) proxy minion and nxos guestshell native minion workflows.  I simply moved this code from the proxy source code file into the module source code file and wanted to preserve existing functionality since I did not add it orginally. 

See: https://github.com/saltstack/salt/pull/49496/files#diff-a9bea6509dcf922bcddad8022f67a474L257

Reply:
Thanks!